### PR TITLE
Retina net

### PR DESCRIPTION
**Pull request by tomkoren21**
Add 2 CaffeLayers and a RetinaNet model testing

### CODE REVIEW

File: test/models/test_retinaNet_decetron2.py [hunk 1]
```diff
@@ -107,3 +105,17 @@ def get_detectron2_models_and_inputs_batch(model_yaml_path:  str = "COCO-Detecti
     tracer = Caffe2Tracer(cfg, model, sample_inputs) #this might raise w
     onnx_model = tracer.export_onnx() #this requires pip install onnx==1.8.1
     return onnx_model, model, sample_inputs
+
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize('change_ordering', [True, False])
```
Comment (doronHarnoy) [inline test/models/test_retinaNet_decetron2.py]: You can set change ordering to False as we use our own `change_ordering` and not the one of onnx2keras

**Comment by tomkoren21** [comment]
Hi  - @doronHarnoy @tip3x.  whoever goes over this  PR - I suggest you ping me before/after having a look since there were some issues with the RetinaNet unit testing.

File: test/models/test_retinaNet_decetron2.py [hunk 2]
```diff
@@ -107,3 +105,17 @@ def get_detectron2_models_and_inputs_batch(model_yaml_path:  str = "COCO-Detecti
...
+@pytest.mark.parametrize('model_class', ["COCO-Detection/retinanet_R_50_FPN_1x.yaml"])
```
Comment (doronHarnoy) [inline test/models/test_retinaNet_decetron2.py]: If they have different trained weights you can parametrize them here also

**Comment by doronHarnoy** [comment]
@tomkoren21 im on it right now 😄

File: test/detectron2_utils.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    bbox_cascade_reg = cfg['MODEL']['ROI_BOX_CASCADE_HEAD']['BBOX_REG_WEIGHTS']
+    bbox_cascade_iou = cfg['MODEL']['ROI_BOX_CASCADE_HEAD']['IOUS']
+    #
+    tracer = Caffe2Tracer(cfg, model, sample_inputs) #this might raise w
+    onnx_model = tracer.export_onnx() #this requires pip install onnx==1.8.1
+    return onnx_model, model, sample_inputs
+
+
+def get_detectron2_models_and_inputs_batch(model_yaml_path:  str = "COCO-Detection/retinanet_R_50_FPN_1x.yaml"):
+    model_yaml = "COCO-Detection/retinanet_R_50_FPN_1x.yaml"
```
Comment (doronHarnoy) [inline test/detectron2_utils.py]: this parameter is the same value as `model_yaml_path` default value, is it intentional?

File: test/models/test_retinaNet_detectron2.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    sample_path = join(pathlib.Path(__file__).parent.parent.resolve(), "data", "detectron2", "test.jpg")
+    sample_inputs = get_sample_inputs(sample_image=sample_path, cfg=cfg, batch=False)
+    model(sample_inputs)
+    Caffe2MetaArch._caffe2_preprocess_image = _caffe2_preprocess_image
+    tracer = Caffe2Tracer(cfg, model, sample_inputs) #this might raise w
+    onnx_model = tracer.export_onnx() #this requires pip install onnx==1.8.1
+    return onnx_model, model, sample_inputs
+
+
+def check_torch_keras_error(model, k_model, input_dict, epsilon=1e-5, change_ordering=False,
```
Comment (doronHarnoy) [inline test/models/test_retinaNet_detectron2.py:90]: we have this function already in our testing no? why did you copy it here?

File: test/models/retinanet/requirements.txt [hunk 5]
```diff
@@ -0,0 +1,89 @@
+absl-py==1.0.0
+antlr4-python3-runtime==4.8
+appdirs==1.4.4
+astunparse==1.6.3
+attrs==21.4.0
+black==22.3.0
+cachetools==5.1.0
+certifi==2022.5.18.1
+charset-normalizer==2.0.12
+click==8.1.3
```
Comment (yotamazriel) [inline test/models/retinanet/requirements.txt:10]: Let's remove this file

File: test/models/retinanet/RETINANET_README.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+To test the RetinaNet you need to install the retinaNet env:
+
+1. Create a virtual environment that uses python 3.8 (via Poetry or Conda) and activate it
+2. pip install -r requirements.txt
+3. python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'
```
Comment (yotamazriel) [inline test/models/retinanet/RETINANET_README.md:5]: It won't work man we need to run it on CI.
Add skip on the test for now and we will get back to it later

**Review by yotamazriel — COMMENTED** [review]
Consider adding only the functions that support that import.
and open a new pr in a draft mode with the tests and Doron or Idan will take it from there

**Comment by tomkoren21** [comment]
Follwing a talk with @doronHarnoy  I'm moving this to draft untill we solve some issues. Will open a seperate PR to the two layers added here.

File: test/models/test_retinaNet_decetron2.py (hunk 2 above)
Comment (tomkoren21) [reply test/models/test_retinaNet_decetron2.py]: no other train weight (for now)

File: test/detectron2_utils.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+import pathlib
+from os.path import join
+from detectron2.modeling import build_model
+from detectron2.checkpoint import DetectionCheckpointer
+
+import onnx
+import pytest
+
+
+def get_sample_inputs(sample_image, cfg, mode='test', batch=False):
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply test/detectron2_utils.py]: fixed

File: test/models/retinanet/requirements.txt (hunk 5 above)
Comment (tomkoren21) [reply test/models/retinanet/requirements.txt:10]: Marked as draft until we resolve this issue

File: test/models/retinanet/RETINANET_README.md (hunk 6 above)
Comment (tomkoren21) [reply test/models/retinanet/RETINANET_README.md:5]: skipped the test and changed PR to draft

File: test/models/test_retinaNet_detectron2.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        error = np.max(np.abs(p - k))
+        np.testing.assert_allclose(p, k, atol=epsilon, rtol=0.0)
+        if error > max_error:
+            max_error = error
+
+    return max_error
+
+
+@pytest.mark.slow
+@pytest.mark.parametrize('change_ordering', [True, False])
```
[the remark this answers is not in the document]
Comment (tomkoren21) [reply test/models/test_retinaNet_detectron2.py:134]: Done

**Pull request closed without merging**